### ISSUE TITLE
Erroneous line allowed to parse, which complains about setup.Cabal.

### ISSUE THREAD

**Issue by chrissound** [haskell/cabal#7018]
**Describe the bug**

With the following cabal file:
```
cabal-version: 1.12
name:                HaskellNixCabalStarter
version:             0.1.0.0
author:              HaskellNixCabalStarter
maintainer:          HaskellNixCabalStarter
license:             MIT
synopsis:       A collection of Attoparsec combinators for parsing org-mode
flavored documents.
build-type:     Simple
category:       Data

executable app
  main-is: Main.hs
  other-modules:
      Paths_HaskellNixCabalStarter
  hs-source-dirs:
      src
  build-depends:
      base >=4.12 && <4.13
  default-language: Haskell2010
```

Which I assume is incorrect... Due to the `flavored documents.` under `synopsis:...` (it should be indented instead). I get the following `cabal v2-repl` output:

```
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: HaskellNixCabalStarter-0.1.0.0 (user goal)
[__1] next goal: HaskellNixCabalStarter:setup.Cabal (dependency of
HaskellNixCabalStarter)
[__1] rejecting: HaskellNixCabalStarter:setup.Cabal-2.4.0.1/installed-2.4...
(conflict: HaskellNixCabalStarter => HaskellNixCabalStarter:setup.Cabal>=1.12
&& <1.25)
[__1] rejecting: HaskellNixCabalStarter:setup.Cabal-3.2.0.0 (constraint from
maximum version of Cabal used by Setup.hs requires <3.2)
[__1] rejecting: HaskellNixCabalStarter:setup.Cabal-3.0.2.0,
HaskellNixCabalStarter:setup.Cabal-3.0.1.0,
HaskellNixCabalStarter:setup.Cabal-3.0.0.0,
HaskellNixCabalStarter:setup.Cabal-2.4.1.0,
HaskellNixCabalStarter:setup.Cabal-2.4.0.1,
HaskellNixCabalStarter:setup.Cabal-2.4.0.0 (conflict: HaskellNixCabalStarter
=> HaskellNixCabalStarter:setup.Cabal>=1.12 && <1.25)
[__1] rejecting: HaskellNixCabalStarter:setup.Cabal-2.2.0.1,
HaskellNixCabalStarter:setup.Cabal-2.2.0.0,
HaskellNixCabalStarter:setup.Cabal-2.0.1.1,
HaskellNixCabalStarter:setup.Cabal-2.0.1.0,
HaskellNixCabalStarter:setup.Cabal-2.0.0.2,
HaskellNixCabalStarter:setup.Cabal-1.24.2.0,
HaskellNixCabalStarter:setup.Cabal-1.24.0.0,
HaskellNixCabalStarter:setup.Cabal-1.22.8.0,
HaskellNixCabalStarter:setup.Cabal-1.22.7.0,
HaskellNixCabalStarter:setup.Cabal-1.22.6.0,
HaskellNixCabalStarter:setup.Cabal-1.22.5.0,
HaskellNixCabalStarter:setup.Cabal-1.22.4.0,
HaskellNixCabalStarter:setup.Cabal-1.22.3.0,
HaskellNixCabalStarter:setup.Cabal-1.22.2.0,
HaskellNixCabalStarter:setup.Cabal-1.22.1.1,
HaskellNixCabalStarter:setup.Cabal-1.22.1.0,
HaskellNixCabalStarter:setup.Cabal-1.22.0.0,
HaskellNixCabalStarter:setup.Cabal-1.20.0.4,
HaskellNixCabalStarter:setup.Cabal-1.20.0.3,
HaskellNixCabalStarter:setup.Cabal-1.20.0.2,
HaskellNixCabalStarter:setup.Cabal-1.20.0.1,
HaskellNixCabalStarter:setup.Cabal-1.20.0.0,
HaskellNixCabalStarter:setup.Cabal-1.18.1.7,
HaskellNixCabalStarter:setup.Cabal-1.18.1.6,
HaskellNixCabalStarter:setup.Cabal-1.18.1.5,
HaskellNixCabalStarter:setup.Cabal-1.18.1.4,
HaskellNixCabalStarter:setup.Cabal-1.18.1.3,
HaskellNixCabalStarter:setup.Cabal-1.18.1.2,
HaskellNixCabalStarter:setup.Cabal-1.18.1.1,
HaskellNixCabalStarter:setup.Cabal-1.18.1,
HaskellNixCabalStarter:setup.Cabal-1.18.0,
HaskellNixCabalStarter:setup.Cabal-1.16.0.3,
HaskellNixCabalStarter:setup.Cabal-1.16.0.2,
HaskellNixCabalStarter:setup.Cabal-1.16.0.1,
HaskellNixCabalStarter:setup.Cabal-1.16.0,
HaskellNixCabalStarter:setup.Cabal-1.14.0,
HaskellNixCabalStarter:setup.Cabal-1.12.0,
HaskellNixCabalStarter:setup.Cabal-1.10.2.0,
HaskellNixCabalStarter:setup.Cabal-1.10.1.0,
HaskellNixCabalStarter:setup.Cabal-1.10.0.0,
HaskellNixCabalStarter:setup.Cabal-1.8.0.6,
HaskellNixCabalStarter:setup.Cabal-1.8.0.4,
HaskellNixCabalStarter:setup.Cabal-1.8.0.2,
HaskellNixCabalStarter:setup.Cabal-1.6.0.3,
HaskellNixCabalStarter:setup.Cabal-1.6.0.2,
HaskellNixCabalStarter:setup.Cabal-1.6.0.1,
HaskellNixCabalStarter:setup.Cabal-1.4.0.2,
HaskellNixCabalStarter:setup.Cabal-1.4.0.1,
HaskellNixCabalStarter:setup.Cabal-1.4.0.0,
HaskellNixCabalStarter:setup.Cabal-1.2.4.0,
HaskellNixCabalStarter:setup.Cabal-1.2.3.0,
HaskellNixCabalStarter:setup.Cabal-1.2.2.0,
HaskellNixCabalStarter:setup.Cabal-1.2.1,
HaskellNixCabalStarter:setup.Cabal-1.1.6,
HaskellNixCabalStarter:setup.Cabal-1.24.1.0 (constraint from minimum version
of Cabal used by Setup.hs requires >=2.4)
[__1] fail (backjumping, conflict set: HaskellNixCabalStarter,
HaskellNixCabalStarter:setup.Cabal)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: HaskellNixCabalStarter:setup.Cabal,
HaskellNixCabalStarter
```

If I remove that line it works correctly... This took a lot of time debugging. 

**Expected behavior**

In my opinion Cabal should instead report this erroneous line as soon as possible. Rather than complain about `setup.Cabal` - (which was also difficult to know what even IS setup.cabal...)

```
cabal-install version 3.0.0.0
compiled using version 3.0.0.0 of the Cabal library 
```

**Comment by phadej** [comment]
After `cabal format`, which outputs warninngs

```
% cabal format
Warning: /home/phadej/mess/issue-7018.cabal/issue-7018.cabal:10:1: Ignoring
trailing fields after sections: "category"
Warning: /home/phadej/mess/issue-7018.cabal/issue-7018.cabal:9:1: Ignoring
trailing fields after sections: "build-type"
Warning: /home/phadej/mess/issue-7018.cabal/issue-7018.cabal:8:1: Ignoring
section: "flavored"
```

the `.cabal` file `cabal` "sees" i.e. uses is

```
cabal-version: 1.12
name:          HaskellNixCabalStarter
version:       0.1.0.0
license:       MIT
maintainer:    HaskellNixCabalStarter
author:        HaskellNixCabalStarter
synopsis:      A collection of Attoparsec combinators for parsing org-mode

executable app
    main-is:          Main.hs
    hs-source-dirs:   src
    other-modules:    Paths_HaskellNixCabalStarter
    default-language: Haskell2010
    build-depends:    base >=4.12 && <4.13
```

---

On original file `cabal check` outputs a lot of issues

```
% cabal check
Warning: The following warnings are likely to affect your build negatively:
Warning: 'hs-source-dirs: src' directory does not exist.
Warning: No 'build-type' specified. If you do not need a custom Setup.hs or
./configure script then use 'build-type: Simple'.
Warning: These warnings may cause trouble when distributing the package:
Warning: No 'category' field.
Warning: No 'description' field.
Warning: A 'license-file' is not specified.
Warning: issue-7018.cabal:10:1: Ignoring trailing fields after sections:
"category"
Warning: issue-7018.cabal:9:1: Ignoring trailing fields after sections:
"build-type"
Warning: issue-7018.cabal:8:1: Ignoring section: "flavored"
Warning: From version 1.24 cabal supports specifying explicit dependencies for
Custom setup scripts. Consider using cabal-version >= 1.24 and adding a
'custom-setup' section with a 'setup-depends' field that specifies the
dependencies of the Setup.hs script itself. The 'setup-depends' field uses the
same syntax as 'build-depends', so a simple example would be 'setup-depends:
base, Cabal'.
Warning: The following errors will cause portability problems on other
environments:
Warning: The filename ./issue-7018.cabal does not match package name
(expected: HaskellNixCabalStarter.cabal)
Warning: The package is missing a Setup.hs or Setup.lhs script.
Warning: Hackage would reject this package.
```

---

TL;DR `flavored documents.` looks like section header (like `executable foo`), and thus all top-level fields are ignored.

**Comment by chrissound** [comment]
This is the first time I'm aware of `cabal format` or `cabal check`. Would it not be useful to include these warnings with `cabal v2-repl`?

**Comment by phadej** [comment]
`v2-repl` is less chatty (when constructing build plan) than other commands (e.g. `v2-build`). That is by design.

`v2-build` would printed the same warnings as `cabal format`.

**Comment by chrissound** [comment]
> That is by design.

I think this design can be improved, because as mentioned I spent nearly an hour or two trying to figure out this issue (and I've been using Haskell for 3+ years now). But I appreciate the help none the less, thanks.

**Comment by phadej** [comment]
@chrissound constructive feedback is welcome.

The basic guideline, that if something doesn't work, and you don't know why, is to increase verbosity level (with `-v` or `-v2`). That would helped in this case too. By default having `cabal` log everything possible would become annoying.
